Mathlib/Data/Nondet/Basic.lean

```diff
@@ -0,0 +1,176 @@
+  L.casesM
+    (fun _ => pure .nil)
+    (fun ⟨s, x⟩ xs => do match ← (do restoreState s; MLList.uncons (f x)) with
+    | none => return bind xs f
+    | some (y, ys) => return .cons y (ys.append (fun _ => bind xs f)))
+
+/-- `Nondet m` is a monad. -/
+instance : Monad (Nondet m) where
+  pure a := .singletonM do return (← saveState, a)
+  bind := bind
+
+/-- `Nondet m` is an alternative monad. -/
+instance : Alternative (Nondet m) where
+  failure := MLList.nil
+  orElse := MLList.append
+
+/-- Convert any value in the monad to the singleton nondeterministic value. -/
+def monadLift (x : m α) : Nondet m α :=
+  .singletonM do
+    let a ← x
+    return (← saveState, a)
+
+instance : MonadLift m (Nondet m) where
+  monadLift := monadLift
+
+/--
+Lift a list of monadic values to a nondeterministic value.
+We ensure that each monadic value is evaluated with the same backtrackable state.
+-/
+def ofListM [Alternative m] (L : List (m α)) : Nondet m α :=
+  MLList.squash fun _ => do
+    let s ← saveState
+    return MLList.ofList L |>.mapM fun x => do
+      restoreState s
+      let a ← x
+      pure (← saveState, a)
+
+/--
+Lift a list of values to a nondeterministic value.
+(The backtrackable state in each will be identical:
+whatever the state was when we first read from the result.)
+-/
+def ofList [Alternative m] (L : List α) : Nondet m α := ofListM (L.map pure)
+
+/--
+Squash a monadic nondeterministic value to a nondeterministic value.
+-/
+def squash (L : Unit → m (Nondet m α)) : Nondet m α := MLList.squash L
+
+/-- Apply a function which returns values in the monad to every alternative of a `Nondet m α`. -/
+partial def mapM (f : α → m β) (L : Nondet m α) : Nondet m β :=
+  L.casesM
+    (fun _ => return .nil)
+    (fun ⟨s, x⟩ xs => do
+      restoreState s
+      let r ← f x
+      let s' ← saveState
+      return .cons (s', r) (Nondet.mapM f xs))
+
+/-- Apply a function to each alternative in a `Nondet m α` . -/
+def map (f : α → β) (L : Nondet m α) : Nondet m β :=
+  L.mapM fun a => pure (f a)
+
+/-- Filter a nondeterministic value using a monadic predicate. -/
+partial def filterM (p : α → m (ULift Bool)) (L : Nondet m α) : Nondet m α :=
+  L.casesM
+    (fun _ => return .nil)
+    (fun ⟨s, x⟩ xs => do
+      restoreState s
+      let r := (← p x).down
+      let s' ← saveState
+      return if r then do
+        .cons (s', x) (filterM p xs)
+      else
+        (filterM p xs))
+
+/-- Filter a nondeterministic value. -/
+def filter (p : α → Bool) (L : Nondet m α) : Nondet m α :=
+  L.filterM fun a => pure <| .up (p a)
+
+/-- Filter and map a nondeterministic value using a monadic function which may return `none`. -/
+partial def filterMapM (f : α → m (Option β)) (L : Nondet m α) : Nondet m β :=
+  L.casesM
+    (fun _ => return .nil)
+    (fun ⟨s, x⟩ xs => do
+      restoreState s
+      match ← f x with
+      | some b => do return .cons (← saveState, b) (filterMapM f xs)
```

Comment:
```suggestion
      | some b => .cons (← saveState, b) (filterMapM f xs)
```
or is there some clever monadlift thing happening here?

Reply:
Obsolete after complete rewrite.